src/preamble.js

```diff
@@ -1669,6 +1669,46 @@ function lookupSymbol(ptr) { // for a pointer, print out all symbols that resolv
 var memoryInitializer = null;
 
 #if USE_PTHREADS
+
+// JS library / EM_ASM proxying
+
+function proxyToMainThread() {
+  var args = Array.prototype.slice.call(arguments);
+  // Arguments are:
+  //   * the index of the function
+  //   * the call args
+  //   * whether it is sync
+  var sync = args[args.length - 1];
+  // The serialization buffer contains the number of call params, and then
+  // all the args here.
+  // We also pass 'sync' to C separately, since C needs to look at it.
+  // The buffer remains alive until receiveOnMainThread frees it.
+  var bufferLen = args.length + 1;
+  var buffer = _malloc(bufferLen * 8); // TODO: stackAlloc if sync?
+  var numCallArgs = args.length - 2;
+  HEAPF64[buffer >> 3] = numCallArgs; // num of call args
```

Comment:
does this need to be `buffer >>> 3`?

Reply:
For 4gb, yes, but for now I'd like to keep it consistent with everything else. I'll deal with 4gb separately.